ui/src/saros/ui/browser/SWTBrowserWrapper.java

```diff
@@ -0,0 +1,192 @@
+          state = ReadyState.COMPLETE;
+        }
+
+        @Override
+        public void changed(ProgressEvent event) {
+          state = ReadyState.LOADING;
+        }
+      };
+
+  public SWTBrowserWrapper(Composite parent, int style) {
+    browser = new Browser(parent, style);
+
+    browser.addDisposeListener(
+        (DisposeEvent e) -> {
+          for (Runnable runnable : runOnDisposalList) {
+            runnable.run();
+          }
+        });
+  }
+
+  @Override
+  public boolean setFocus() {
+    return this.syncExec(
+        () -> {
+          return browser.setFocus();
+        });
+  }
+
+  @Override
+  public void setSize(final int width, final int height) {
+    this.syncExec(() -> browser.setSize(width, height));
+  }
+
+  @Override
+  public String getUrl() {
+    return this.syncExec(
+        () -> {
+          return browser.getUrl();
+        });
+  }
+
+  @Override
+  public boolean loadUrl(final String url, final long timeout) {
+    browser.getShell().open();
+
+    this.state = ReadyState.LOADING;
+    browser.addProgressListener(readyStateListener);
+
+    browser.setUrl(url);
+    boolean successful = this.waitForPageLoaded(url, timeout);
+
+    browser.removeProgressListener(readyStateListener);
+
+    return successful;
+  }
+
+  /**
+   * Waits for a page to be loaded.
+   *
+   * @param url
+   * @param timeout
+   * @return
+   */
+  private boolean waitForPageLoaded(final String url, final long timeout) {
+
+    Runnable waitForLoaded =
+        () -> {
+          while (this.state != ReadyState.COMPLETE) {
+            if (!browser.getDisplay().readAndDispatch()) {
+              browser.getDisplay().sleep();
+            }
+          }
+        };
+
+    Thread t = new Thread(waitForLoaded);
+    t.setDaemon(true);
+    t.start();
+
+    try {
+      t.join(timeout);
+    } catch (InterruptedException e) {
+      return false;
+    }
+    return true;
+  }
+
+  @Override
+  public void loadHtml(final String html) {
+    this.syncExec(() -> browser.setText(html));
+  }
+
+  @Override
+  public Object evaluate(final String jsCode) {
+    return this.syncExec(
+        () -> {
+          Object rv = browser.evaluate(jsCode);
+          System.out.println(rv);
+          return rv;
+        });
+  }
+
+  @Override
+  public boolean execute(String jsCode) {
+    return this.syncExec(
+        () -> {
+          return browser.execute(jsCode);
+        });
+  }
+
+  @Override
+  public void addBrowserFunction(final JavascriptFunction function) {
+    this.syncExec(
+        () -> {
+          BrowserFunction swtFunction =
+              new BrowserFunction(browser, function.getName()) {
+                @Override
+                public Object function(Object[] arguments) {
+                  return function.function(arguments);
+                }
+              };
+          return swtFunction;
+        });
+  }
+
+  @Override
+  public void runOnDisposal(Runnable runnable) {
```

Comment:
Race conditions. As it seems that this class can be used from any thread this list is accessed by multiple threads and ArrayLists are not thread safe.

Reply:
I will adjust it